Lambda layer release automation

reakaleek opened the pull request:
## What does this pull request do?

Based on https://github.com/elastic/apm-agent-python/pull/1796

Relevant commit range: https://github.com/elastic/apm-agent-python/pull/1814/files/766da7e9b5e9f8c4d5067ffac3f4c4087ade5554..436c7f5c8748bfb7fffac092bd763e963dc2f3ac

- Adds lambda layer release automation
- Add creation of github release draft to facilitate the arn file

<!-- Comment:
Here you can explain the changes made on the PR.
-->

## How to test

```
ELASTIC_LAYER_NAME=test ./.ci/publish-aws.sh
```

With valid AWS credentials

reakaleek commented on .github/workflows/release.yml:
```diff
@@ -39,9 +41,56 @@ jobs:
+          path: ".arn-file.md"
+          if-no-files-found: error
+
+  github-draft:
+    needs:
+      - publish-lambda-layers
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/download-artifact@v3
+        with:
+          name: arn-file
+      - run: >- 
+          gh release create "${GITHUB_REF_NAME}"
+          --title="${GITHUB_REF_NAME}"
+          --generate-notes
+          --notes-file=".arn-file.md"
+        env:
+          GH_TOKEN: ${{ github.token }}
```
@basepi 

I'm creating a github release draft here, which can be adjusted and published manually later.

I did this so we have the arn table already in the draft.

<details>
<summary> example arn table created during testing </summary>

### Elastic APM Python agent layer ARNs

|Region|ARN|
|------|---|
|ap-northeast-1|arn:aws:lambda:ap-northeast-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ap-northeast-2|arn:aws:lambda:ap-northeast-2:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ap-northeast-3|arn:aws:lambda:ap-northeast-3:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ap-south-1|arn:aws:lambda:ap-south-1:627286350134:layer:jcal-dev-elastic-apm-python:8|
|ap-southeast-1|arn:aws:lambda:ap-southeast-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ap-southeast-2|arn:aws:lambda:ap-southeast-2:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ca-central-1|arn:aws:lambda:ca-central-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-central-1|arn:aws:lambda:eu-central-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-north-1|arn:aws:lambda:eu-north-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-west-1|arn:aws:lambda:eu-west-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-west-2|arn:aws:lambda:eu-west-2:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-west-3|arn:aws:lambda:eu-west-3:627286350134:layer:jcal-dev-elastic-apm-python:6|
|sa-east-1|arn:aws:lambda:sa-east-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|us-east-1|arn:aws:lambda:us-east-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|us-east-2|arn:aws:lambda:us-east-2:627286350134:layer:jcal-dev-elastic-apm-python:6|
|us-west-1|arn:aws:lambda:us-west-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|us-west-2|arn:aws:lambda:us-west-2:627286350134:layer:jcal-dev-elastic-apm-python:6|



</details>

reakaleek commented on .github/workflows/release.yml:
```diff
@@ -39,9 +41,57 @@ jobs:
+          path: ".arn-file.md"
+          if-no-files-found: error
+
+  github-draft:
+    needs:
+      - publish-lambda-layers
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/download-artifact@v3
+        with:
+          name: arn-file
+      - run: >- 
+          gh release create "${GITHUB_REF_NAME}"
+          --title="${GITHUB_REF_NAME}"
+          --generate-notes
+          --notes-file=".arn-file.md"
+          --draft
+        env:
+          GH_TOKEN: ${{ github.token }}
```
@basepi 

I'm creating a github release draft here, which can be adjusted and published manually later.

I did this so we have the arn table already in the draft.

<details>
<summary> example arn table created during testing </summary>

### Elastic APM Python agent layer ARNs

|Region|ARN|
|------|---|
|ap-northeast-1|arn:aws:lambda:ap-northeast-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ap-northeast-2|arn:aws:lambda:ap-northeast-2:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ap-northeast-3|arn:aws:lambda:ap-northeast-3:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ap-south-1|arn:aws:lambda:ap-south-1:627286350134:layer:jcal-dev-elastic-apm-python:8|
|ap-southeast-1|arn:aws:lambda:ap-southeast-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ap-southeast-2|arn:aws:lambda:ap-southeast-2:627286350134:layer:jcal-dev-elastic-apm-python:6|
|ca-central-1|arn:aws:lambda:ca-central-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-central-1|arn:aws:lambda:eu-central-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-north-1|arn:aws:lambda:eu-north-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-west-1|arn:aws:lambda:eu-west-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-west-2|arn:aws:lambda:eu-west-2:627286350134:layer:jcal-dev-elastic-apm-python:6|
|eu-west-3|arn:aws:lambda:eu-west-3:627286350134:layer:jcal-dev-elastic-apm-python:6|
|sa-east-1|arn:aws:lambda:sa-east-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|us-east-1|arn:aws:lambda:us-east-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|us-east-2|arn:aws:lambda:us-east-2:627286350134:layer:jcal-dev-elastic-apm-python:6|
|us-west-1|arn:aws:lambda:us-west-1:627286350134:layer:jcal-dev-elastic-apm-python:6|
|us-west-2|arn:aws:lambda:us-west-2:627286350134:layer:jcal-dev-elastic-apm-python:6|

</details>


WDYT?

basepi replied:
Looks great!

reakaleek commented:
@elastic/apm-agent-python 

Not sure if related, but the windows tests seem to be stuck.

amannocci commented on .github/workflows/release.yml:
```diff
@@ -39,9 +44,66 @@ jobs:
+  publish-lambda-layers:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: hashicorp/vault-action@v2.4.2
```
```suggestion
      - uses: hashicorp/vault-action@v2.5.0
```
Use the latest version available.

amannocci commented on .github/workflows/release.yml:
```diff
@@ -39,9 +44,66 @@ jobs:
+          roleId: ${{ secrets.VAULT_ROLE_ID }}
+          secretId: ${{ secrets.VAULT_SECRET_ID }}
+          secrets: |
+            secret/observability-team/ci/service-account/apm-agent-python access_key_id | AWS_ACCESS_KEY_ID ;
```
We could use an env variable to avoid duplication.

reakaleek replied:
Can you elaborate?

amannocci replied:
```suggestion
            ${{ env.SECRET_PATH }} access_key_id | AWS_ACCESS_KEY_ID ;
```
We could define an env variable on this step and expand it in the secret input.
**Note:** It's a small cosmetic

reakaleek replied:
9e2fe41a

amannocci reviewed (approved):
LGTM

**Note:** Some minor feedbacks

v1v commented on .ci/publish-aws.sh:
```diff
@@ -0,0 +1,59 @@
+    --region="${region}" \
+    --layer-name="${FULL_LAYER_NAME}" \
+    --description="AWS Lambda Extension Layer for the Elastic APM Python Agent" \
+    --license="Apache-2.0" \
```
As far as I see the `apm-agent-python` uses a different [license](https://github.com/elastic/apm-agent-python/blob/main/LICENSE). But I don't know what's the one for the AWS lambda, can we clarify this?

reakaleek replied:
Good point. It's copied from apm-aws-lambda. I mistakenly assumed it will be the same.

reakaleek replied:
Any suggestions @basepi @beniwohli ?

reakaleek replied:
https://github.com/elastic/apm-agent-python/commit/30729a05b33e39568457de56bcccc12a29fe681c

v1v commented on .github/workflows/release.yml:
```diff
@@ -39,9 +44,66 @@ jobs:
+      - uses: actions/download-artifact@v3
+        with:
+          name: arn-file
+      - run: >- 
```
NIT: Use name for the step?

reakaleek replied:
```suggestion
      - name: Create GitHub Draft Release
        run: >- 
```

v1v commented on .github/workflows/release.yml:
```diff
@@ -39,9 +44,66 @@ jobs:
+    if: always()
+    needs:
+      - publish-lambda-layers
+      - release
```
What about `github-draft`? Though, it has a dependency with `publish-lambda-layers`, if added, will it work in the case `github-draft` failed?

reakaleek replied:
```suggestion
      - publish-pypi
      - github-draft
```

v1v reviewed (commented):
LGTM, with some comments

basepi reviewed (approved):
Looking great! Thank you for the fast turnaround here.

reakaleek commented:
> Looking great! Thank you for the fast turnaround here. Once we fix up the license and the other review feedback we should be good.

https://github.com/elastic/apm-agent-python/pull/1814/commits/30729a05b33e39568457de56bcccc12a29fe681c

reakaleek commented:
@basepi one last thing that I noticed while checking the license.

what are the compatible runtimes?

these are the available options:

```
  python2.7
  python3.6
  python3.7
  python3.8
```

basepi commented:
> what are the compatible runtimes?

3.6, 3.7, 3.8, 3.9 (3.9 is an option too, it's just up at the top of the dropdown with the other "latest versions")

reakaleek commented:
> > what are the compatible runtimes?
> 
> 3.6, 3.7, 3.8, 3.9 (3.9 is an option too, it's just up at the top of the dropdown with the other "latest versions")

Ah ups, thanks for double checking. I was looking at the example of the cli documentation

https://awscli.amazonaws.com/v2/documentation/api/2.0.34/reference/lambda/publish-layer-version.html and the list is incomplete.

reakaleek commented:
> 3.6, 3.7, 3.8, 3.9 (3.9 is an option too, it's just up at the top of the dropdown with the other "latest versions")

https://github.com/elastic/apm-agent-python/pull/1814/commits/18f561b26b787b135a970a5c5507fd2f93ba8a16

v1v commented on .github/workflows/release.yml:
```diff
@@ -5,12 +5,17 @@ on:
     tags:
       - "v*.*.*"
 
+permissions:
```
NIT: if it is only used when creating the draft release,  then it can be configured in that particular job rather than globally

reakaleek replied:
https://github.com/elastic/apm-agent-python/pull/1814/commits/3510ffd61185615e72e8726397e26a5fed77a1a7

v1v reviewed (approved):
LGTM, with a minor comment

basepi commented:
Windows tests are unrelated: https://github.com/elastic/apm-agent-python/pull/1817

reakaleek commented:
@basepi 

To be sure, are we good to merge this PR including the changes from https://github.com/elastic/apm-agent-python/pull/1796?

basepi commented:
> @reakaleek are we good to merge this PR including the changes from https://github.com/elastic/apm-agent-python/pull/1796?

Please do not merge this. #1796 needs to be reviewed and merged first, then we'll merge the automation.

basepi commented:
@reakaleek I updated https://github.com/elastic/apm-agent-python/pull/1796 to better handle the docker image paths, and added a Dockerfile. You'll want to rebase.

Similar to the [nodejs docker file](https://github.com/elastic/apm-agent-nodejs/blob/main/Dockerfile), it expects an `AGENT_DIR` arg, which in the this case should be `./build/dist/package/python`.

This will result in the same directory structure in both the lambda layer (mounted at `/opt`) and the docker image.

Can you please add support for building and uploading the docker image? Thanks!

reakaleek commented:
> @reakaleek I updated #1796 to better handle the docker image paths, and added a Dockerfile. You'll want to rebase.
> 
> Similar to the [nodejs docker file](https://github.com/elastic/apm-agent-nodejs/blob/main/Dockerfile), it expects an `AGENT_DIR` arg, which in the this case should be `./build/dist/package/python`.
> 
> This will result in the same directory structure in both the lambda layer (mounted at `/opt`) and the docker image.
> 
> Can you please add support for building and uploading the docker image? Thanks!

https://github.com/elastic/apm-agent-python/pull/1814/commits/63b7294653a70cb88203470504df828a302e15e9

basepi commented:
https://github.com/elastic/apm-agent-python/pull/1796 is merged, so you can rebase and merge this at your convenience.

basepi commented:
Actually, I have one more request, @reakaleek -- I think we should be testing `make-distribution.sh` to make sure we don't end up with a dependency conflict in `dev-utils/requirements.txt`. Since I'm [dynamically injecting into requirements.txt](https://github.com/elastic/apm-agent-python/blob/1e38ec53115edc70c36c6485259733a8cde02ed9/dev-utils/make-distribution.sh#L41) at build time, Dependabot can't properly check for incompatibilities in the dependency chain.

I could throw a `set -e` at the beginning of the script and then write a unit test that runs this command. But I think it probably makes sense to run it as a github action. Do you agree?

reakaleek commented:
> Actually, I have one more request, @reakaleek -- I think we should be testing `make-distribution.sh` to make sure we don't end up with a dependency conflict in `dev-utils/requirements.txt`. Since I'm [dynamically injecting into requirements.txt](https://github.com/elastic/apm-agent-python/blob/1e38ec53115edc70c36c6485259733a8cde02ed9/dev-utils/make-distribution.sh#L41) at build time, Dependabot can't properly check for incompatibilities in the dependency chain. I think we should be testing this regularly, rather than finding out it fails when we try to release.
> 
> I could throw a `set -e` at the beginning of the script and then write a unit test that runs this command. But I think it probably makes sense to run it as a github action. Do you agree?
> 
> Edit: We already do `set -o errexit` which is equivalent to `set -e` so we just need to run it and check the retcode.

Hi @basepi,

Sorry for the delay, just got back from sick leave.

> I think we should be testing this regularly, rather than finding out it fails when we try to release.

I agree, testing this regularly makes sense.

> I could throw a set -e at the beginning of the script and then write a unit test that runs this command. But I think it probably makes sense to run it as a github action. Do you agree?

Can you elaborate on this more?  Do you mean we run the script `make-distribution.sh` on a PR level in it's own GH workflow job? Or do you mean we should create a custom action that does it?

reakaleek commented:
> 2f2253e

I extracted the build-distribution script into a reusable workflow and it's now called in the release and the test workflow.

Let me know if this matches your suggestion/request.

https://github.com/elastic/apm-agent-python/pull/1814/commits/2f2253e9b35735ef065d7afdf101c34e382fe828

reakaleek commented:
> > Actually, I have one more request, @reakaleek -- I think we should be testing `make-distribution.sh` to make sure we don't end up with a dependency conflict in `dev-utils/requirements.txt`. Since I'm [dynamically injecting into requirements.txt](https://github.com/elastic/apm-agent-python/blob/1e38ec53115edc70c36c6485259733a8cde02ed9/dev-utils/make-distribution.sh#L41) at build time, Dependabot can't properly check for incompatibilities in the dependency chain. I think we should be testing this regularly, rather than finding out it fails when we try to release.
> > I could throw a `set -e` at the beginning of the script and then write a unit test that runs this command. But I think it probably makes sense to run it as a github action. Do you agree?
> > Edit: We already do `set -o errexit` which is equivalent to `set -e` so we just need to run it and check the retcode.
> 
> Hi @basepi,
> 
> Sorry for the delay, just got back from sick leave.
> 
> > I think we should be testing this regularly, rather than finding out it fails when we try to release.
> 
> I agree, testing this regularly makes sense.
> 
> > I could throw a set -e at the beginning of the script and then write a unit test that runs this command. But I think it probably makes sense to run it as a github action. Do you agree?
> 
> Can you elaborate on this more? Do you mean we run the script `make-distribution.sh` on a PR level in it's own GH workflow job? Or do you mean we should create a custom action that does it?

I extracted the build-distribution script into a reusable workflow and it's now called in the release and the test workflow.

Let me know if this matches your suggestion/request.

https://github.com/elastic/apm-agent-python/pull/1814/commits/2f2253e9b35735ef065d7afdf101c34e382fe828

basepi commented:
> I extracted the build-distribution script into a reusable workflow and it's now called in the release and the test workflow.
>
> Let me know if this matches your suggestion/request.

This should be perfect! Thank you.
